### PR TITLE
RequestAttributes implementation

### PR DESCRIPTION
**Pull request by flqw**
#12

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -118,6 +118,11 @@
         </dependency>
         <dependency>
             <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-configuration-processor</artifactId>
```
Comment (fiunchinho) [inline pom.xml]: I think this dependency should be added by the clients of this library, not by the library itself. What do you think?

File: src/main/java/net/rakugakibox/springbootext/logback/access/tomcat/RequestAttributeRegardingAccessEvent.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    public String getRemoteAddr() {
+        if (remoteAddress == null) {
+            HttpServletRequest request = getRequest();
+            if (request != null) {
+                remoteAddress = (String) request.getAttribute(AccessLog.REMOTE_ADDR_ATTRIBUTE);
+                if (remoteAddress == null) {
+                    remoteAddress = request.getRemoteAddr();
+                }
+            } else {
+                remoteAddress = NA;
```
Comment (fiunchinho) [inline src/main/java/net/rakugakibox/springbootext/logback/access/tomcat/RequestAttributeRegardingAccessEvent.java]: What's `NA`?

File: pom.xml (hunk 1 above)
Comment (flqw) [inline pom.xml]: This needs to be here as it generates a meta-data file that will be packaged it into the jar of the library. As an example, this enables auto-completion in the properties in IntelliJ Idea.

File: src/main/java/net/rakugakibox/springbootext/logback/access/tomcat/RequestAttributeRegardingAccessEvent.java (hunk 2 above)
Comment (flqw) [inline src/main/java/net/rakugakibox/springbootext/logback/access/tomcat/RequestAttributeRegardingAccessEvent.java]: A static variable that indicates that the value is not present (coming from the IAccessEvent interface).

File: pom.xml (hunk 1 above)
Comment (fiunchinho) [inline pom.xml]: Yes, but if all it does is to generate auto-completion in the properties in an specific IDE, I'd say that the decision about whether this is required or not belongs to the end user of this library.

Comment (flqw) [inline pom.xml]: This is not IDE-specific, that was just an example. The meta-data files are present in all spring libraries and this is an official feature. See [here](http://docs.spring.io/spring-boot/docs/current/reference/html/configuration-metadata.html#configuration-metadata-annotation-processor).

Comment (flqw) [inline pom.xml]: Also it is not possible to generate metadata files for third party libraries. This can't be a user decision.

Comment (fiunchinho) [inline pom.xml]: I see what you mean now. Ok, lets leave it like this. But if I understand this Spring feature correctly, this is only to get auto-completion and hints when you use these properties?

Comment (flqw) [inline pom.xml]: Yes this is just to help developers.

Comment (akihyro) [inline pom.xml]: I didn't know `spring-boot-configuration-processor`.
Thanks :+1:

File: src/main/java/net/rakugakibox/springbootext/logback/access/LogbackAccessConfigurator.java [hunk 3]
```diff
@@ -36,13 +39,42 @@
...
+        private Boolean enableRequestAttributes;
+
+        public boolean getOrDeduceEnableRequestAttributes() {
+            if (this.enableRequestAttributes != null) {
+                return this.enableRequestAttributes;
+            } else {
+                // Deduce the value from the presence of the RemoteIpValve.
+                this.enableRequestAttributes = servletFactory.getValves()
+                        .stream()
+                        .anyMatch(valve -> RemoteIpValve.class.isAssignableFrom(valve.getClass()));
```
Comment (akihyro) [inline src/main/java/net/rakugakibox/springbootext/logback/access/LogbackAccessConfigurator.java]: cool :+1:

File: src/main/java/net/rakugakibox/springbootext/logback/access/tomcat/RequestAttributeRegardingAccessEvent.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+    private String remoteAddress;
+    private String remoteHost;
+    private String protocol;
+    private Integer localPort;
+
+    public RequestAttributeRegardingAccessEvent(HttpServletRequest httpRequest, HttpServletResponse httpResponse,
+                                                ServerAdapter adapter) {
+        super(httpRequest, httpResponse, adapter);
+        prepareForDeferredProcessing();
```
Comment (akihyro) [inline src/main/java/net/rakugakibox/springbootext/logback/access/tomcat/RequestAttributeRegardingAccessEvent.java]: Can I ask you why call `prepareForDeferredProcessing()`?
I think it is better to be lazy, like the parent class.
I passed the test with it.

File: src/test/java/net/rakugakibox/springbootext/logback/access/tomcat/TomcatRequestAttributesTest.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+ */
+@RunWith(SpringJUnit4ClassRunner.class)
+@SpringApplicationConfiguration
+@WebIntegrationTest(
+        value = {"logback.access.config=classpath:logback-access-test.singleton-queue.xml",
+                "server.use-forward-headers=true",
+                // The Proto and Port Header must be explicitly set to work.
+                "server.tomcat.protocol-header=X-Forwarded-Proto",
+                "server.tomcat.port-header=X-Forwarded-Port",
+                "server.tomcat.protocol-header-https-value=https"
```
Comment (akihyro) [inline src/test/java/net/rakugakibox/springbootext/logback/access/tomcat/TomcatRequestAttributesTest.java]: Can I ask you why set explicitly?
I think it is better to use default value.
I passed the test with it.

* https://github.com/spring-projects/spring-boot/blob/v1.3.5.RELEASE/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/ServerProperties.java#L783-L784
* https://github.com/spring-projects/spring-boot/blob/v1.3.5.RELEASE/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/ServerProperties.java#L583
* https://github.com/spring-projects/spring-boot/blob/v1.3.5.RELEASE/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/ServerProperties.java#L578

File: src/test/java/net/rakugakibox/springbootext/logback/access/test/AccessEventAssert.java [hunk 6]
```diff
@@ -140,6 +140,20 @@ public S hasRemoteAddr(String remoteAddr) {
...
     /**
+     * Verifies that the remote address is equal to the given one.
+     *
+     * @param localPort the remote address.
+     * @return this instance.
+     */
+    public S hasLocalPort(int localPort) {
+        int actualLocalPort = actual.getLocalPort();
+        Assertions.assertThat(actualLocalPort)
+                .isNotEqualTo(NA)
```
Comment (akihyro) [inline src/test/java/net/rakugakibox/springbootext/logback/access/test/AccessEventAssert.java]: Can you change the parameter to `SENTINEL`?
`getLocalPort()` returns `SENTINEL`.

https://github.com/akihyro/spring-boot-ext-logback-access/pull/14/files#diff-4ae93438483ccfa280689788bd60af41R84

Comment (flqw) [inline src/test/java/net/rakugakibox/springbootext/logback/access/test/AccessEventAssert.java]: Yes i will, javadocs seem to be also wrong.

File: src/main/java/net/rakugakibox/springbootext/logback/access/tomcat/RequestAttributeRegardingAccessEvent.java (hunk 4 above)
Comment (flqw) [inline src/main/java/net/rakugakibox/springbootext/logback/access/tomcat/RequestAttributeRegardingAccessEvent.java]: I'll have to look into that again. I think it is a left-over from testing where i had some problems with an asynchronous logger (logback-to-logstash). Console did work fine anyway.

File: src/test/java/net/rakugakibox/springbootext/logback/access/tomcat/TomcatRequestAttributesTest.java (hunk 5 above)
Comment (flqw) [inline src/test/java/net/rakugakibox/springbootext/logback/access/tomcat/TomcatRequestAttributesTest.java]: Which ones did you test without? I think at least the Proto header must be set as it has no default value per http://docs.spring.io/spring-boot/docs/current/reference/html/common-application-properties.html

Comment (akihyro) [inline src/test/java/net/rakugakibox/springbootext/logback/access/tomcat/TomcatRequestAttributesTest.java]: I removed `server.tomcat.*` settings as follows, and then I tested.
The result passed test.

```java
@WebIntegrationTest(
        value = {"logback.access.config=classpath:logback-access-test.singleton-queue.xml",
                "server.use-forward-headers=true",
//                // The Proto and Port Header must be explicitly set to work.
//                "server.tomcat.protocol-header=X-Forwarded-Proto",
//                "server.tomcat.port-header=X-Forwarded-Port",
//                "server.tomcat.protocol-header-https-value=https"
        },
        randomPort = true)
```

But, I was able to understand: `protocol-header` does not have the default value, in the official document.
Let's leave `server.tomcat.*` settings.

**Comment by flqw** [comment]
Okay fixed the open stuff

**Comment by akihyro** [comment]
I merged this PR.
I'm going to release as soon!

**Pull request merged**